### PR TITLE
Expose collector models (configuration) via an API

### PR DESCRIPTION
**Pull request by leenagupte**
The views added:
* Provider View
* DataSource View
* CollectorType View
* Collector View

Each view inherits from Resource View and has a schema.

Move common function calls to base ResourceView class
* Moved call to decorator that prohibits caching of results to Resource View.
* Move the call to the add an Authorization header to Resource View.
* Move finding permissions by resource (model) name to ResourceView

The changes will apply to all operations of all views that inherit from Resource View.

https://www.pivotaltracker.com/story/show/99687256

### CODE REVIEW

File: stagecraft/apps/collectors/views.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+    id_fields = {
+        'id': UUID_RE_STRING,
+        'slug': '[\w-]+',
+    }
+    list_filters = {
+        "name": "name__iexact"
+    }
+
+    def update_model(self, model, model_json, request, parent):
```
Comment (jcbashdown) [inline stagecraft/apps/collectors/views.py]: Dry this up?

**Comment by jcbashdown** [comment]
:+1: Just that small bit of deduping to do.

File: stagecraft/apps/collectors/tests/test_views.py [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+            content_type='application/json')
+
+        assert_that(response.status_code, equal_to(200))
+
+        resp_json = json.loads(response.content)
+
+        assert_that(resp_json['id'], equal_to(str(data_source.id)))
+        assert_that(resp_json['name'], equal_to(data_source.name))
+        assert_that(
+            resp_json['provider']['name'], equal_to(data_source.provider.name))
```
Comment (jcbashdown) [inline stagecraft/apps/collectors/tests/test_views.py]: Check owners not exposed? Check slug is?

File: stagecraft/apps/collectors/tests/test_views.py [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+
+        resp_json = json.loads(resp.content)
+
+        assert_that(resp_json, has_key('id'))
+        assert_that(resp_json['name'], equal_to('some-data-source'))
+        assert_that(resp_json['slug'], equal_to('some-data-source'))
+        assert_that(resp_json['provider']['name'], equal_to(
+            self.provider.name))
+
+    def test_post_from_unauthorised_client_fails(self):
```
Comment (jcbashdown) [inline stagecraft/apps/collectors/tests/test_views.py]: There doesn't seem to be a test in test resource. I would feel a little less paranoid if we can test that non owners can't edit data sources. Something like https://github.com/alphagov/stagecraft/blob/master/stagecraft/apps/dashboards/tests/views/test_dashboard.py#L390

Comment (jcbashdown) [inline stagecraft/apps/collectors/tests/test_views.py]: I suppose maybe also the same for viewing data sources? That's how the ownership array works IIRC?

File: stagecraft/apps/collectors/tests/test_views.py [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+            collector.type.name))
+        assert_that(
+            resp_json['data_source']['name'], equal_to(
+                collector.data_source.name))
+        assert_that(resp_json['data_set']['data_type'], equal_to(
+            collector.data_set.data_type.name))
+        assert_that(resp_json['data_set']['data_group'], equal_to(
+            collector.data_set.data_group.name))
+        assert_that(resp_json['query'], equal_to(collector.query))
+        assert_that(resp_json['options'], equal_to(collector.options))
```
Comment (jcbashdown) [inline stagecraft/apps/collectors/tests/test_views.py]: Maybe also assert owners not exposed here.

**Comment by jcbashdown** [comment]
Just a couple of things because I got paranoid. They may be silly though.

**Comment by leenagupte** [comment]
@jcbashdown I've made the changes you suggested.  Thanks.

**Pull request merged**